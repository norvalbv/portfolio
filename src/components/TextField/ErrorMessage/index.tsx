import React, { ReactElement } from 'react';
import { classNames } from 'utils';

interface ErrorMessageProps {
  message: string;
  dataAtt?: string;
  className?: string;
}

const ErrorMessage = ({ message, dataAtt, className }: ErrorMessageProps): ReactElement => {
  return (
    <div className={classNames(className)}>
      <span
        className="text-red-R500 block max-w-full text-sm"
        data-testid={`${dataAtt ? `${dataAtt} ${message}` : message}`}
      >
        {message}
      </span>
    </div>
  );
};

export default ErrorMessage;
