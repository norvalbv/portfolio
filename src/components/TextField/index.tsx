import { ReactElement, MouseEventHandler, useState, ChangeEvent, useEffect, useRef } from 'react';
import { classNames } from 'utils';
import ErrorMessage from './ErrorMessage';

interface RefObject<T> {
  readonly current: T | null;
}

export interface TextFieldProps {
  /**
   * Autocomplete for the input / textarea element
   *
   * Allow any other string elements:
   * Chrome respects autocomplete=off only when there is at least one other input
   * element in the form with any other autocomplete value.
   *
   * https://stackoverflow.com/a/30873633/17442963
   *
   * On of any string
   */
  autoComplete?: string;
  /**
   * Label for input
   */
  label?: string;
  /**
   * Append component beside the label
   */
  labelAppend?: JSX.Element;
  /**
   * TextField Error
   */
  error?: string | null;
  /**
   * Error message position
   */
  errorMessagePosition?: 'top' | 'bottom';
  /**
   * Type of input
   */
  type?: 'text' | 'email' | 'password';
  /**
   * Input name
   */
  name?: string;
  /**
   * SVG icon
   */
  icon?: JSX.Element;
  /**
   * Position of icon
   */
  iconPosition?: 'left' | 'right';
  /**
   * Input placeholder
   */
  placeholder?: string;
  /**
   * Wrapper classname
   */
  className?: string;
  /**
   * Value of input
   */
  value?: string | number;
  /**
   * Append component beside input
   */
  children?: JSX.Element | JSX.Element[];
  /**
   * Disable input
   */
  disabled?: boolean;
  /**
   * Make field required
   */
  required?: boolean;
  /**
   * Define input id
   */
  id?: string;
  /**
   * Data attribute for testing
   */
  dataAtt?: string;
  /**
   * Optional input event
   */
  onChange?: (value: string) => void;
  /**
   * Optional click input event
   */
  onClick?: MouseEventHandler<HTMLDivElement>;
  /**
   * Input Text Type
   */
  inputType?: 'textarea' | 'input';
  /**
   * Automatically focuses the element
   */
  autoFocus?: boolean;
}

const TextField = ({
  autoComplete = 'on',
  autoFocus = false,
  children,
  className,
  dataAtt,
  disabled,
  error,
  errorMessagePosition = 'bottom',
  icon,
  iconPosition = 'left',
  id,
  inputType = 'input',
  label,
  labelAppend,
  name,
  onChange,
  onClick,
  placeholder,
  required = false,
  type = 'text',
  value,
}: TextFieldProps): ReactElement => {
  const handleChange = (e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>): void => {
    if (!onChange || disabled) {
      return;
    }

    onChange(e.target.value);
  };

  // For when the element is active.
  const [active, setActive] = useState(false);

  const ref = useRef<HTMLInputElement | HTMLTextAreaElement>(null);

  useEffect(() => {
    if (ref.current && autoFocus) {
      ref.current.focus();
    }
  }, [autoFocus]);

  return (
    <fieldset className={className}>
      {label && (
        <div className="mb-2 flex justify-between">
          <div className="text-light-text dark:text-dark-text">
            <label className="flex" htmlFor={id}>
              {label}
              {required && <span className="ml-1 text-xs text-accent-secondary">*</span>}
            </label>
            {labelAppend && <div className="ml-4">{labelAppend}</div>}
          </div>
          {error && errorMessagePosition === 'top' && (
            <ErrorMessage message={error} dataAtt={dataAtt} />
          )}
        </div>
      )}
      <div onClick={onClick}>
        {inputType === 'input' ? (
          <div className="flex items-center">
            <input
              autoComplete={autoComplete}
              className={classNames(
                'appearance-none rounded focus:text-accent-secondary focus:outline focus:outline-light-text focus:ring-transparent dark:focus:outline-dark-text',
                'borber-b h-12 w-full bg-dark-dark/20 p-4 transition-[height] duration-300 dark:bg-white/20'
              )}
              data-testid={dataAtt}
              disabled={disabled}
              id={id}
              name={name || ''}
              onChange={handleChange}
              placeholder={placeholder || ''}
              ref={ref as RefObject<HTMLInputElement>}
              required={required}
              style={{ fontSize: 'inherit' }}
              type={type}
              value={value || ''}
            />
            {iconPosition === 'right' ? icon : children}
          </div>
        ) : (
          <textarea
            autoComplete={autoComplete}
            className={classNames(
              'appearance-none rounded focus:text-accent-secondary focus:outline focus:outline-light-text focus:ring-transparent dark:focus:outline-dark-text',
              'borber-b w-full bg-dark-dark/20 p-4 transition-[height] duration-300 dark:bg-white/20',
              active || value ? 'h-52' : 'h-12'
            )}
            data-testid={dataAtt}
            disabled={disabled}
            id={id}
            name={name || ''}
            onClick={(): void => setActive(true)}
            onBlur={(): void => {
              setTimeout(() => {
                setActive(false);
              }, 150);
            }}
            onChange={handleChange}
            placeholder={placeholder || ''}
            ref={ref as RefObject<HTMLTextAreaElement>}
            required={required}
            value={value || ''}
          />
        )}
      </div>
      {error && errorMessagePosition === 'bottom' && (
        <ErrorMessage message={error} dataAtt={dataAtt} className="mt-2" />
      )}
    </fieldset>
  );
};

export default React.memo(TextField);
