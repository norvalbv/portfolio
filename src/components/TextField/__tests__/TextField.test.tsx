import { FC } from 'react';
import renderer from 'react-test-renderer';

import TextField, { TextFieldProps } from '..';

const TextFieldWithAllProps: FC<Required<TextFieldProps>> = TextField;

describe('<TextField />', () => {
  test('should render correctly with mandatory props', () => {
    const tree = renderer.create(<TextField />);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  test('should render correctly with all props', () => {
    const tree = renderer.create(
      <TextFieldWithAllProps
        autoComplete="on"
        autoFocus={false}
        className="underline"
        dataAtt="Test DataAtt"
        disabled
        error="ERROR!!"
        errorMessagePosition="bottom"
        icon={<div>Icon</div>}
        iconPosition="right"
        id="Test Id"
        inputType="textarea"
        label="Test Label"
        labelAppend={<div>Label</div>}
        name="Test name"
        onChange={jest.fn()}
        onClick={jest.fn()}
        placeholder="placeholder"
        required
        type="text"
        value={30}
      >
        <div>Child</div>
      </TextFieldWithAllProps>
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
