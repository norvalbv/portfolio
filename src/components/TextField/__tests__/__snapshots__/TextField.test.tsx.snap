// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> should render correctly with all props 1`] = `
<fieldset
  className="underline"
>
  <div
    className="mb-2 flex justify-between"
  >
    <div
      className="text-light-text dark:text-dark-text"
    >
      <label
        className="flex"
        htmlFor="Test Id"
      >
        Test Label
        <span
          className="ml-1 text-xs text-accent-secondary"
        >
          *
        </span>
      </label>
      <div
        className="ml-4"
      >
        <div>
          Label
        </div>
      </div>
    </div>
  </div>
  <div
    onClick={[MockFunction]}
  >
    <textarea
      autoComplete="on"
      className="appearance-none rounded focus:text-accent-secondary focus:outline focus:outline-light-text focus:ring-transparent dark:focus:outline-dark-text borber-b w-full bg-dark-dark/20 p-4 transition-[height] duration-300 dark:bg-white/20 h-52"
      data-testid="Test DataAtt"
      disabled={true}
      id="Test Id"
      name="Test name"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      placeholder="placeholder"
      required={true}
      value={30}
    />
  </div>
  <div
    className="mt-2"
  >
    <span
      className="block max-w-full text-sm text-red-500"
      data-testid="Test DataAtt ERROR!!"
    >
      ERROR!!
    </span>
  </div>
</fieldset>
`;

exports[`<TextField /> should render correctly with mandatory props 1`] = `
<fieldset>
  <div>
    <div
      className="flex items-center"
    >
      <input
        autoComplete="on"
        className="appearance-none rounded focus:text-accent-secondary focus:outline focus:outline-light-text focus:ring-transparent dark:focus:outline-dark-text borber-b h-12 w-full bg-dark-dark/20 p-4 transition-[height] duration-300 dark:bg-white/20"
        name=""
        onChange={[Function]}
        placeholder=""
        required={false}
        style={
          {
            "fontSize": "inherit",
          }
        }
        type="text"
        value=""
      />
    </div>
  </div>
</fieldset>
`;
